containers-int-strict/src/Map/Int.hs

```diff
@@ -32,61 +35,62 @@ module Map.Int
        ) where
 
 import Prelude hiding (lookup, null)
+import Data.Coerce (coerce)
 
 import qualified Data.IntMap.Strict as M
 
-newtype Map k v = IM { unIM :: M.IntMap v }
+newtype Map k v = IM (M.IntMap v)
     deriving newtype (Show, Eq)
 
 type Key = (~) Int
 
-empty :: Map k v
-empty = IM M.empty
+empty :: forall k v. Map k v
+empty = coerce @(M.IntMap v) M.empty
 
-singleton :: Key k => k -> v -> Map k v
-singleton k = IM . M.singleton k
+singleton :: forall k v. Key k => k -> v -> Map k v
+singleton = coerce @(k -> v -> M.IntMap v) M.singleton
 
-fromList :: Key k => [(k, v)] -> Map k v
-fromList = IM . M.fromList
+fromList :: forall k v. Key k => [(k, v)] -> Map k v
+fromList = coerce @([(k, v)] -> M.IntMap v) M.fromList
 
-null :: Map k v -> Bool
-null = M.null . unIM
+null :: forall k v. Map k v -> Bool
+null = coerce @(M.IntMap v -> Bool) M.null
 
-size :: Map k v -> Int
-size = M.size . unIM
+size :: forall k v. Map k v -> Int
+size = coerce @(M.IntMap v -> Int) M.size
 
-member :: Key k => k -> Map k a -> Bool
-member k = M.member k . unIM
+member :: forall k v. Key k => k -> Map k v -> Bool
+member = coerce @(k -> M.IntMap v -> Bool) M.member
 
-lookup :: Key k => k -> Map k v -> Maybe v
-lookup k = M.lookup k . unIM
+lookup :: forall k v. Key k => k -> Map k v -> Maybe v
+lookup = coerce @(k -> M.IntMap v -> Maybe v) M.lookup
 
-lookupDefault :: Key k => v -> k -> Map k v -> v
-lookupDefault def k = M.findWithDefault def k . unIM
+lookupDefault :: forall k v. Key k => v -> k -> Map k v -> v
+lookupDefault = coerce @(v -> k -> M.IntMap v -> v) M.findWithDefault
 
-toList :: Key k => Map k v -> [(k, v)]
-toList = M.toList . unIM
+toList :: forall k v. Key k => Map k v -> [(k, v)]
+toList = coerce @(M.IntMap v -> [(k, v)]) M.toList
 
-keys :: Key k => Map k v -> [k]
-keys = M.keys . unIM
+keys :: forall k v. Key k => Map k v -> [k]
+keys = coerce @(M.IntMap v -> [k]) M.keys
 
-elems :: Map k v -> [v]
-elems = M.elems . unIM
+elems :: forall k v. Map k v -> [v]
+elems = coerce @(M.IntMap v -> [v]) M.elems
 
-insert :: Key k => k -> v -> Map k v -> Map k v
-insert k v = IM . M.insert k v . unIM
+insert :: forall k v. Key k => k -> v -> Map k v -> Map k v
+insert = coerce @(k -> v -> M.IntMap v -> M.IntMap v) M.insert
 
-insertWith :: Key k => (v -> v -> v) -> k -> v -> Map k v -> Map k v
-insertWith f k v= IM . M.insertWith f k v . unIM
+insertWith :: forall k v. Key k => (v -> v -> v) -> k -> v -> Map k v -> Map k v
+insertWith = coerce @((v -> v -> v) -> k -> v -> M.IntMap v -> M.IntMap v) M.insertWith
```

Comment:
Yeah, the implementation is awkward... Fortunately, we need to write it only once :joy:

Reply:
Just don't look under the hood :joy: